src/main/scala/ai/tripl/arc/plugins/pipeline/ControlFlowExecute.scala

```diff
@@ -0,0 +1,162 @@
+
+        val stage = ControlFlowExecuteStage(
+          plugin=this,
+          name=name,
+          key=key,
+          description=description,
+          inputURI=parsedURI,
+          sql=inputSQL,
+          sqlParams=sqlParams,
+          params=params
+        )
+
+        stage.stageDetail.put("key", key)
+        stage.stageDetail.put("inputURI", parsedURI.toString)
+        stage.stageDetail.put("sql", inputSQL)
+        stage.stageDetail.put("sqlParams", sqlParams.asJava)
+        stage.stageDetail.put("params", params.asJava)
+
+        Right(stage)
+      case _ =>
+        val allErrors: Errors = List(name, key, description, parsedURI, inputSQL, validSQL, invalidKeys).collect{ case Left(errs) => errs }.flatten
+        val stageName = stringOrDefault(name, "unnamed stage")
+        val err = StageError(index, stageName, c.origin.lineNumber, allErrors)
+        Left(err :: Nil)
+    }
+  }
+}
+
+case class ControlFlowExecuteStage(
+    plugin: ControlFlowExecute,
+    name: String,
+    key: String,
+    description: Option[String],
+    inputURI: URI,
+    sql: String,
+    sqlParams: Map[String, String],
+    params: Map[String, String]
+  ) extends PipelineStage {
+
+  override def execute()(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+    ControlFlowExecuteStage.execute(this)
+  }
+}
+
+case class ControlFlowPayload(
+  outcome: Boolean,
+  message: String
+)
+
+object ControlFlowExecuteStage {
+
+  def execute(stage: ControlFlowExecuteStage)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
+
+    val signature = "ControlFlowExecute requires query to return 1 row with [outcome: boolean, message: string] signature."
+
+    val stmt = SQLUtils.injectParameters(stage.sql, stage.sqlParams, false)
+    stage.stageDetail.put("sql", stmt)
+
+    val df = try {
+      spark.sql(stmt)
+    } catch {
+      case e: Exception => throw new Exception(e) with DetailException {
+        override val detail = stage.stageDetail
+      }
+    }
+    val count = df.persist(arcContext.storageLevel).count
+
+    if (df.count != 1 || df.schema.length != 2) {
+      throw new Exception(s"""${signature} Query returned ${count} rows of type [${df.schema.map(f => f.dataType.simpleString).mkString(", ")}].""") with DetailException {
+        override val detail = stage.stageDetail
+      }
+    }
+
+
+    try {
+      val row = df.first
+      val resultIsNull = row.isNullAt(0)
+      val messageIsNull = row.isNullAt(1)
+
+      if (resultIsNull) {
+        throw new Exception(s"""${signature} Query returned [null, ${if (messageIsNull) "null" else "not null"}].""") with DetailException {
+          override val detail = stage.stageDetail
+        }
+      }
+
+      val message = row.getString(1)
+
+      // try to parse to json
+      try {
```

Comment:
This code has been repeated so consider making it a utility method.

Reply:
by moving the outcome to ControlFlowPayload this logic is not repeated in ControlFlowExecute anymore. It is still repeated in SQLValidate but not sure that matters for such small code.